### ISSUE TITLE
2019: HuggingFace's Transformers: State-of-the-art Natural Language Processing

### ISSUE THREAD

**Issue by jojonki** [jojonki/arXivNotes#295]
HuggingFace's Transformers: State-of-the-art Natural Language Processing
Thomas Wolf, Lysandre Debut, Victor Sanh, Julien Chaumond, Clement Delangue, Anthony Moi, Pierric Cistac, Tim Rault, Rémi Louf, Morgan Funtowicz, Jamie Brew
11 pages, 1 figure, more details at this https URL
https://arxiv.org/abs/1910.03771
code: https://github.com/huggingface/transformers

## 概要
HuggingFace社のTransformerライブラリの紹介論文．PyTorchとTFでモデルは実装されており，様々なツールやモデルが出揃っている．NLPシステムを作るためのスタンダードなライブラリとなるべく開発を頑張っている．2019年10月の段階でTransformerのスター数は15.4Kと大人気

## イントロ
ここ18ヶ月では，多くのNLPタスクが巨大なデータで事前学習された言語モデルによる転移学習が利用されている．Transformerを使った方式は，テキスト分類，言語理解，機械翻訳，ゼロショット言語生成，共参照解決，常識推論など他分野に及ぶ．

事前学習型は素晴らしい性能を示した一方，計算資源の点で問題がある．Facebook AIのRoBERTa（Liu 2019)は，V100 (32GB RAM)を1024枚使って，160GBのデータを学習した．AWSを利用してこれをやろうとすると，100K USDかかる．

このようなトレンドとは対象的に，機械学習やNLPのコミュニティでは知識共有のためコードやモデルの共有が推進されている．これにより追試が簡単になり，新しい仮説の実験が行いやすくなる．

- 簡単アクセスでハイパフォーマンス
Transofmersは2つのゴールがある．// 論文ではTransformersはライブラリ名として記載されている
1) 可能な限り簡単で早い，2) オリジナル論文で報告されてる性能に近い値を出す．ライブラリの学習コストは非常に低く，`from_pretrained()`という関数を使って簡単にインスタンス化できる．

- 解釈性と多様性
[BERTology](https://huggingface.co/transformers/bertology.html)と呼ばれるBERTの内部動作の分析に関する事例も含まれている．そのために1) 隠れ状態，アテンション重み，ヘッド重要性などへのアクセスIF，2) 特定のアーキテクチャへの過学習を避けるために異なるモデルに対する統合API，を提供．更にGLUE, SuperGLUE, SQuAD1.1といったデータセットの事前処理及びfine-tuningスクリプトもついている．

- ベストプラクティスの推奨
PyTorch及びTFに対応

- リサーチからプロダクション
産業界の人のために，リサーチからプロダクションへのスムーズな遷移も狙っている．PyTorchからシリアライズと最適化を行うTorchScriptの提供，TensorFlow Extendedの統合も行っている．

## コミュニティ
もともとはHuggingfaceのオープンソースの内部ツールだったが，だんだん人気が出てきて，pytorch-pretrained-bert，pytorch-transformers, Transformersとリポジトリの名前を変えていった．リサーチと産業界で人気が出ており，すでに200以上の論文で本ライブラリが使われている．またTransformersはSpacy, AllenNLP, Flairなどにも依存性やラッパーとして含まれている．

Transoformersは進行中のPjであり，120の外部貢献者もいる．Apache 2.0で利用でき，pipやGithubから利用が可能だ．Apache 2.0なので無料，商用，改変，再配布，アプリ同梱などが可能で比較的会社にも緩いライセンス．（Swiftもそうらしい）

## ライブラリデザイン
3つの基本コンポーネントについて以下に説明していく: 1. configuration，2. model, 3. tokenization．全コンポーネントはPyTorchとTensorFlow互換．

### 主要コンポーネント
- Configuration
設定クラスはPretrainedConfigクラスを継承したもので，モデル，トーカナイザのパラメタ（語彙サイズ，隠れ次元数，ドロップアウト率）を保存している．設定オブジェクトは再現性のため保存，ロードされる．この設定は，モデルのアーキテクチャだけでなく，headの枝刈りのようなアーキ最適化も定義している．またDNNフレームワークに囚われずこの設定クラスは利用できる．

- Tokenizers
トーカナイザクラスはPreTrainedTokenizerクラスを継承したもので，各モデルで利用できる．語彙toIndexを持ち，入力シーケンスのエンコーディング・デコーディングに対応（BPE, SentencrePieceなど）．またユーザー定義のトークンを追加したり，語彙サイズのリサイズなどできる．
Tokenizerクラスは，事前定義モデルのTransformersで定義済みのものからinstantiateされたり，ユーザー定義のものから作られたりする．

- Model
モデルのベースクラスは，エンコーディング（embedding層のプロジェクション）からセルフアテンション，最終レイヤの隠れ層に至る計算グラフを実装している．ベースクラスは，モデルのオリジナルの実装をできるだけ再現実装している．更にユーザーは内部の実装状態もかんたんに紐解いて分析することができる．
ベースクラスのトップには追加のラパークラスがあり，ベースモデルの隠れ状態に追加できる．例えば言語モデルやシーケンス分類ヘッドなどである．これらのクラスは，XXXForSequenceClassification, XXXForMaskedLMといった名前があり，fine-tuningや事前学習に利用できる（XXXはモデル名）

すべてのモデルはPyTorchとTFで利用でき，標準のシリアライゼーションライブラリを使ってモデルを保存することで，PyTorch⇔TFで利用できる．

- 自動クラス
Autoクラスでは，モデル名やパスを受け取って，該当するクラスをinstantiateする．AutoClassでは異なるモデル・モデル・トーカナイザ間で高速なスイッチンを可能にする統合APIを提供する．４つのハイレベル抽象クラスが提供されており，AutoConfig, AutoTokenizer, AutoModel (PyTorch), TFAutoModel．これらのクラスは自動で，事前定義のチェックポイント名（bert-base-cased等）から，正しい設定，トーカナイザ，モデルクラスをinstantiateする

### 学習
- Optimizer
PyTorchのOptimizerのサブクラスとして最適化ユーティリティを提供し，AdamやAdamWを提供

- Scheduler
PyTorchのLambdaLRのサブクラスとして提供され，学習率のスケジューラーが利用可能

## Transformersで実験
言語理解と言語生成の実験で本ライブラリを使ってみよう

### 言語理解ベンチマーク
GLUE, SuperGLUEを評価するツールとスクリプトを用意．機械理解ではSQuAD 1.1, 2.0．またSWAG, RACE, ARCなども追加されている．

### 言語モデルのfine-tuning
Transormersはユーザー独自のテキストセットで簡単にfine-tuningできる．またトークンの追加・削除もできる

### エコシステム

- [Write With Transformer](https://transformer.huggingface.co)
Write with TransformerはGPT, GPT2, XLNetといったアーキを簡単にWebブラウザ上で試せるインタフェースである．
// なんかしらテキストが与えた状態でTabキーを教えていくと単語が数個ずつ自動で生成されていって面白い
// 下記はXLNetで生成したもの（太字部分が生成部）
![image](https://user-images.githubusercontent.com/166852/67154472-1d7d9b80-f338-11e9-94ad-195440d0e0cb.png)

- [Conversational AI](https://github.com/huggingface/transfer-learning-conv-ai)
HuggingFaceではe2eで言語理解とテキスト生成を行う対話エージェントTalking Dogを実装しており，2018秋にデモをしており，[Conversational Intelligence Challenge 2](http://convai.io/)（NIPS2018)でSOTAに到達するアプローチだった．コードやライブラリはgithubから利用可能．ConvAI2では第２位の性能の模様．PersonChatと呼ばれる雑談対話を与えられ，自動評価と人評価で性能がでるチャレンジの模様．

- プロダクションでの利用
Research -> Productionへのスムーズな遷移のために，TorchScriptを用意．これはPyTorchモデルの中間表現を生成し，PythonやC++などで実行できるようになる．

プロダクションのために大きな機械学習モデルを最適化することは進行中で最も注目を浴びている領域であり，DistilBERT（これもHuggingFace）のように性能を維持しつつモデルを最適化して小さく早くする活動が活発．これにより遅延制約が厳しい環境や安価なCPUサーバーでの実行を可能にする．またモバイル向けのモデル変換コードも提供している（https://github.com/kaushaltrivedi/fast-bert）

> 参考：オリジナルより60%高速/軽量で、ELMoより概ね高性能+オリジナルの95%程度の性能を達成。
// https://github.com/arXivTimes/arXivTimes/issues/1414

- Community
多くのNLPライブラリがTransformers上に作られている．確認できるところで，FastBert, FARM, Flair, AllenNLP, PyText．

## アーキテクチャ
Transformersによって提供されているのは下記の通り．生成モデル（GPT, GPT-2, Transformer-XL, XLNet, XLM）と言語理解モデル（Bert, DistilBert, RoBERTa，XLM）が有る．

いくつ句のモデルではSQuAD1.1などのダウンストリームタスクにfine-tuneしたモデルも利用可能．結果として，30の事前学習の重みが利用可能であり，10以上の英語以外の言語で提供されている．multi-lingualモデル（100言語以上で学習されたものも含む）

## 関連研究
略

## Conclusion
Transformersを紹介した．ResearchでもIndustryでも簡単に様々な事前学習モデルを利用できる．そしてすでに広く使われている．高いソフトウェアエンジニアリングと機械学習エンジニアリングを維持しつつ，アクセスが良く簡単に使える転移学習の開発環境をこれからも作っていきたい．
